File: rel-grid-workbench/src/main/resources/homing/js/hue/captains/singapura/js/homing/relgrid/workbench/GamesStore.js
// =============================================================================
// GamesStore — the Games Catalogue bench's store: seven hundred-odd releases,
// 1990 to 2020, read from the catalogue's pipe-separated text. DOMAIN CODE;
// the word "grid" does not appear in it. Read-only: the catalogue is a fact,
// and what this bench exercises is ordering and choosing, not editing.
//
//   createGamesStore(psv)          the catalogue: a header line, then a row a line, '|'-separated;
//                                  lines that start with # are remarks
//   store.pks()                    'g001', 'g002', … in the catalogue's order
//   store.columns()                the header's names, in order
//   store.kind(col)                'text' | 'number' | 'set' — how a column's values compare and are chosen
//   store.get(pk, col)             the value: a number for a number column, null where the
//                                  catalogue left it blank; a string otherwise
//   store.distinct(col)            a set column's values with their counts, most frequent first
//   store.size()
//
// Numbers are numbers here — year, sales, score — so a sort compares them as
// numbers and a filter ranges over them; a blank is null, deliberately, so a
// sort has to say what it does with absence rather than sorting '' as text.
// =============================================================================

var _WB_GAMES_KINDS = { year: "number", sales: "number", score: "number", platform: "set", type: "set" };

function createGamesStore(psv) {
    var lines = String(psv || "").split(/\r?\n/), columns = null, rows = new Map(), pks = [];
    for (var k = 0; k < lines.length; k++) {
        var line = lines[k];
        if (!line || line.charAt(0) === "#") continue;
        var parts = line.split("|");
        if (!columns) { columns = parts.slice(1); continue; }             // the header: 'id' and then the columns
        var pk = parts[0], row = {};
        for (var c = 0; c < columns.length; c++) {
            var raw = parts[c + 1] === undefined ? "" : parts[c + 1];
            row[columns[c]] = (_WB_GAMES_KINDS[columns[c]] === "number") ? (raw === "" ? null : Number(raw)) : raw;
        }
        rows.set(pk, row);
        pks.push(pk);
    }
    columns = columns || [];
    var distinct = {};

    return {
        pks:     function () { return pks.slice(); },
        columns: function () { return columns.slice(); },
        kind:    function (col) { return _WB_GAMES_KINDS[col] || "text"; },
        get:     function (pk, col) { var r = rows.get(pk); return r ? r[col] : undefined; },
        size:    function () { return pks.length; },
        distinct: function (col) {
            if (!distinct[col]) {
                var counts = new Map();
                for (var i = 0; i < pks.length; i++) { var v = rows.get(pks[i])[col]; counts.set(v, (counts.get(v) || 0) + 1); }
                var out = [];
                counts.forEach(function (n, v) { out.push({ value: v, count: n }); });
                out.sort(function (a, b) { return b.count - a.count || String(a.value).localeCompare(String(b.value)); });
                distinct[col] = out;
            }
            return distinct[col].slice();
        }
    };
}

File: rel-grid-workbench/src/main/java/hue/captains/singapura/js/homing/relgrid/workbench/GamesStore.java
package hue.captains.singapura.js.homing.relgrid.workbench;

import hue.captains.singapura.js.homing.core.EsModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;

import java.util.List;

/**
 * The Games Catalogue's store: seven hundred-odd releases, 1990 to 2020, read
 * from the catalogue's pipe-separated text. Read-only, and pure logic — no
 * DOM, no persistence: a fact the bench orders and chooses from. Numbers are
 * numbers here and a blank is null, so a sort compares as numbers and has
 * to say what it does with absence.
 */
public record GamesStore() implements EsModule<GamesStore> {

    public record createGamesStore() implements Exportable._Constant<GamesStore> {}

    public static final GamesStore INSTANCE = new GamesStore();

    @Override public ImportsFor<GamesStore> imports() { return ImportsFor.noImports(); }

    @Override public ExportsOf<GamesStore> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new createGamesStore()));
    }
}
